cassandra-commons/src/main/java/com/mesosphere/dcos/cassandra/common/tasks/CassandraDaemonTask.java

```diff
@@ -414,206 +63,161 @@ private final CassandraConfig updateConfig(CassandraDaemonStatus status) {
-        try {
-            return CassandraProtos.CassandraTaskData.newBuilder()
-                    .setType(
-                            CassandraProtos.CassandraTaskData.TYPE.CASSANDRA_DAEMON)
-                    .setConfig(config.toProto())
-                    .setAddress(hostname).build();
-        } catch (IOException ex) {
-            throw new RuntimeException(ex);
-        }
+    public CassandraMode getMode() {
+        return getData().getMode();
     }
 
     @Override
     public CassandraDaemonTask update(CassandraTaskStatus status) {
-
         if (status.getType() == TYPE.CASSANDRA_DAEMON &&
-                status.getId().equals(id)) {
-
-            return create(
-                    id,
-                    slaveId,
-                    hostname,
-                    executor,
-                    name,
-                    role,
-                    principal,
-                    cpus,
-                    memoryMb,
-                    diskMb,
-                    updateConfig((CassandraDaemonStatus) status),
-                    (CassandraDaemonStatus) status);
+            status.getId().equals(getId())) {
+            CassandraDaemonStatus daemonStatus = (CassandraDaemonStatus) status;
+            return new CassandraDaemonTask(
+                Protos.TaskInfo.newBuilder(getTaskInfo())
+                    .setData(getData().updateDaemon(
+                        daemonStatus.getState(),
+                        daemonStatus.getMode(),
+                        updateConfig(daemonStatus.getState(), getConfig())
+                    ).getBytes()).build()
+            );
         }
         return this;
 
     }
 
+    public VolumeRequirement.VolumeType getVolumeType() {
+        return getConfig().getDiskType();
+    }
+
     @Override
     public CassandraDaemonTask update(Protos.TaskState state) {
+        return new CassandraDaemonTask(
+            getBuilder()
+                .setData(getData()
+                    .withState(state)
+                    .getBytes()).build());
 
+    }
 
-        return create(
-                id,
-                slaveId,
-                hostname,
-                executor,
-                name,
-                role,
-                principal,
-                cpus,
-                memoryMb,
-                diskMb,
-                config,
-                ((CassandraDaemonStatus) status).update(state));
+    public CassandraDaemonStatus createStatus(Protos.TaskState state,
+                                              CassandraMode mode,
+                                              Optional<String> message) {
+        return CassandraDaemonStatus.create(getStatusBuilder(state, message)
+            .setData(CassandraData.createDaemonStatusData(mode).getBytes())
+            .build());
+    }
 
+    public String getVolumePath() {
+        return TaskUtils.getVolumePaths(
+            getTaskInfo().getResourcesList())
+            .get(0);
```

Comment:
I want us to re-think about this. Just getting `0` always might lead us into trouble, if we ever want to support multiple volumes.